### DISCUSSION TITLE
ba.peakhold

### DISCUSSION THREAD

**Discussion by dariosanfilippo** [grame-cncm/faustlibraries#126] [General]
Hello, people.

I was testing Faust's `ba.peakholder` and I have a few doubts about how it works.

The description says "tracks abs peak and holds peak for 'n' samples" but it appears that the timer is not reset when a new peak is detected; is that the expected behaviour?

The mechanism that I'd expect from a peak holder is that when a new peak is detected the timer is reset, otherwise the timer is started. Hence, the impulse response of an N-sample hold period is an N + 1 step.

It could be up for debate whether an input that is equal to the output is considered a new peak or not, and I personally think that it should be.

See the test code below.

Ciao,
Dario

```
import("stdfaust.lib");
peakHold(t, x) = loop ~ si.bus(2) : ! , _
    with {
        loop(timerState, outState) = timer , output
            with {
                isNewPeak = abs(x) >= outState;
                isTimeOut = timerState >= rint(t * ma.SR);
                bypass = isNewPeak | isTimeOut;
                timer = (1 - bypass) * (timerState + 1);
                output = bypass * (abs(x) - outState) + outState;
            };
    };
dirac = 1-1';
process = (dirac + (dirac @ 20) * 1.1 + (dirac @ 40) * 1.2 + (dirac @ 60) * 1.3) <: _ , peakHold(.001) , ba.peakholder(rint(.001 * ma.SR));
```

**Comment by sletz** [comment]
Code written by [@lij](https://github.com/lijon) (Jonatan Liljedahl  on Discord) hopefully he can help here?

**Comment by lijon** [comment]
My own version of this code is a tiny bit different, but not sure if it makes a difference in your case. Looking at the block diagram I see that the intention is that the timer should only run while the held value is larger than the input value. When the input is greater than the held value, the timer (sweep) should reset. This is the behaviour you're asking for if I understand correctly? I haven't tested (recently at least) if it actually behaves as intended..

https://faustide.grame.fr/?autorun=0&voices=0&name=untitled&inline=cGVha2hvbGQgPSAoKixfOm1heCkgfiBfOwpzd2VlcCA9ICUoaW50KCo6bWF4KDEpKSl-KygxKTsKcGVha2hvbGRlcihob2xkdGltZSkgPSBwZWFraG9sZF9nYXRlIH4gcmVzZXQgOiAoISxfKSB3aXRoIHsKICAgIHJlc2V0ID0gc3dlZXAoaG9sZHRpbWUpID4gMDsKICAgIC8vIGZpcnN0IG91dCBpcyBnYXRlIHRoYXQgaXMgMSB3aGlsZSBob2xkaW5nIGxhc3QgcGVhawogICAgcGVha2hvbGRfZ2F0ZSA9IF8sXyA8OiBwZWFraG9sZCwhLF8gPDogPj0sXywhOwp9Owpwcm9jZXNzID0gcGVha2hvbGRlcigxMDApOw%3D%3D
